File: cpp_version/src/http/http_parser.cpp
#include "http/http_parser.h"

HttpParser::HttpParser() : state_(HttpParserState::METHOD) {}

void HttpParser::reset() {
    state_ = HttpParserState::METHOD;
    request_ = HttpRequest();
    current_header_key_.clear();
}

bool HttpParser::parse(const char* data, size_t len) {
    for (size_t i = 0; i < len; ++i) {
        char c = data[i];
        switch (state_) {
            case HttpParserState::METHOD:
                if (c == ' ') {
                    state_ = HttpParserState::URL;
                } else {
                    request_.method += c;
                }
                break;
            case HttpParserState::URL:
                if (c == ' ') {
                    state_ = HttpParserState::PROTOCOL;
                } else {
                    request_.url += c;
                }
                break;
            case HttpParserState::PROTOCOL:
                if (c == '\r') {
                    // Ignore
                } else if (c == '\n') {
                    state_ = HttpParserState::HEADER_KEY;
                } else {
                    request_.protocol += c;
                }
                break;
            case HttpParserState::HEADER_KEY:
                if (c == ':') {
                    state_ = HttpParserState::HEADER_VALUE;
                } else if (c == '\r') {
                    // Ignore
                } else if (c == '\n') {
                    state_ = HttpParserState::BODY;
                } else {
                    current_header_key_ += c;
                }
                break;
            case HttpParserState::HEADER_VALUE:
                if (c == '\r') {
                    // Ignore
                } else if (c == '\n') {
                    request_.headers[current_header_key_] = current_header_key_;
                    current_header_key_.clear();
                    state_ = HttpParserState::HEADER_KEY;
                } else if (c != ' ' || !current_header_key_.empty()) {
                    current_header_key_ += c;
                }
                break;
            case HttpParserState::BODY:
                request_.body += c;
                break;
        }
    }
    return true;
}

File: rust_version/src/http/http_parser.rs
use std::collections::HashMap;

enum HttpParserState {
    Method,
    Url,
    Protocol,
    HeaderKey,
    HeaderValue,
    Body,
}

struct HttpRequest {
    method: String,
    url: String,
    protocol: String,
    headers: HashMap<String, String>,
    body: String,
}

struct HttpParser {
    state: HttpParserState,
    request: HttpRequest,
    current_header_key: String,
}

impl HttpParser {
    fn new() -> Self {
        HttpParser {
            state: HttpParserState::Method,
            request: HttpRequest {
                method: String::new(),
                url: String::new(),
                protocol: String::new(),
                headers: HashMap::new(),
                body: String::new(),
            },
            current_header_key: String::new(),
        }
    }

    fn reset(&mut self) {
        self.state = HttpParserState::Method;
        self.request = HttpRequest {
            method: String::new(),
            url: String::new(),
            protocol: String::new(),
            headers: HashMap::new(),
            body: String::new(),
        };
        self.current_header_key.clear();
    }

    fn parse(&mut self, data: &str) -> bool {
        for c in data.chars() {
            match self.state {
                HttpParserState::Method => {
                    if c == ' ' {
                        self.state = HttpParserState::Url;
                    } else {
                        self.request.method.push(c);
                    }
                }
                HttpParserState::Url => {
                    if c == ' ' {
                        self.state = HttpParserState::Protocol;
                    } else {
                        self.request.url.push(c);
                    }
                }
                HttpParserState::Protocol => {
                    if c == '\r' {
                        // Ignore
                    } else if c == '\n' {
                        self.state = HttpParserState::HeaderKey;
                    } else {
                        self.request.protocol.push(c);
                    }
                }
                HttpParserState::HeaderKey => {
                    if c == ':' {
                        self.state = HttpParserState::HeaderValue;
                    } else if c == '\r' {
                        // Ignore
                    } else if c == '\n' {
                        self.state = HttpParserState::Body;
                    } else {
                        self.current_header_key.push(c);
                    }
                }
                HttpParserState::HeaderValue => {
                    if c == '\r' {
                        // Ignore
                    } else if c == '\n' {
                        if !self.current_header_key.is_empty() {
                            self.request.headers.insert(self.current_header_key.clone(), String::new());
                            self.current_header_key.clear();
                        }
                        self.state = HttpParserState::HeaderKey;
                    } else {
                        if self.current_header_key.is_empty() {
                            // First character of value, start capturing
                            self.request.headers.insert(self.current_header_key.clone(), c.to_string());
                        } else {
                            if let Some(value) = self.request.headers.get_mut(&self.current_header_key) {
                                value.push(c);
                            }
                        }
                    }
                }
                HttpParserState::Body => {
                    self.request.body.push(c);
                }
            }
        }
        true
    }
}
